scl/rewrite/cc-mask.conf

```diff
@@ -38,22 +38,63 @@
 
-@define balabit.credit-card-regexp "(:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})"
+block rewrite credit-card-internal(regex() value() template())
+{
+   subst("`regex`"
+         "`template`"
+         value(`value`)
+         flags(global, store-matches)
+         type(pcre)
+    );
+};
+
+block rewrite credit-card-visa(value() template())
+{
+  credit-card-internal(regex("(?:4[0-9]{12}(?:[0-9]{3})?)") value(`value`) template(`template`));
+};
+
+block rewrite credit-card-mastercard(value() template())
+{
+  credit-card-internal(regex("((5[1-5][0-9]{14})|((?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}))") value(`value`) template(`template`));
+};
+
+block rewrite credit-card-american_express(value() template())
+{
+  credit-card-internal(regex("(3[47][0-9]{13})") value(`value`) template(`template`));
+};
+
+block rewrite credit-card-diners_club(value() template())
+{
+  credit-card-internal(regex("(3(?:0[0-5]|[68][0-9])[0-9]{11})") value(`value`) template(`template`));
+};
+
+block rewrite credit-card-discover(value() template())
+{
+  credit-card-internal(regex("(6(?:011|5[0-9]{2})[0-9]{12})") value(`value`) template(`template`));
+};
+
+block rewrite credit-card-jcb(value() template())
+{
+  credit-card-internal(regex("((?:2131|1800|35[0-9]{3})[0-9]{11})") value(`value`) template(`template`));
+};
+
 
-block rewrite credit-card-hash(value("MESSAGE"))
+block rewrite credit-card-mask(value("MESSAGE") template("$(substr $0 0 6)******$(substr $0 -4 4)"))
 {
-        subst("`balabit.credit-card-regexp`",
-              "$(sha1 --length 16 $1)",
-              value(`value`),
-              flags(global, store-matches),
-              type(pcre));
+    credit-card-visa(value(`value`) template("`template`"));
+    credit-card-mastercard(value(`value`) template("`template`"));
+    credit-card-american_express(value(`value`) template("`template`"));
+    credit-card-jcb(value(`value`) template("`template`"));
+    credit-card-diners_club(value(`value`) template("`template`"));
+    credit-card-discover(value(`value`) template("`template`"));
 };
 
-block rewrite credit-card-mask(value("MESSAGE"))
+block rewrite credit-card-hash(value("MESSAGE") template("$(sha1 --length 16 $0)"))
 {
-        subst("`balabit.credit-card-regexp`",
-              "$(substr $1 0 6)******$(substr $1 -4 4)"
-              value(`value`),
-              flags(global, store-matches),
-              type(pcre));
+    credit-card-visa(value(`value`) template("`template`"));
+    credit-card-mastercard(value(`value`) template("`template`"));
+    credit-card-american_express(value(`value`) template("`template`"));
+    credit-card-jcb(value(`value`) template("`template`"));
+    credit-card-diners_club(value(`value`) template("`template`"));
+    credit-card-discover(value(`value`) template("`template`"));
 };
```

Comment:
This is easier to read but at the same time definitely slower than the original implementation where we had a single regexp, while now we have 7 of them. I can live with it, I just wanted to note it.

Reply:
Note: Talked about this on the last OSE meeting and we decided to keep this. We are not even 100% sure, that the new implementation is slower.